interface/ceed-basis.c

```diff
@@ -1085,6 +1085,29 @@ int CeedBasisApply(CeedBasis basis, CeedInt num_elem, CeedTransposeMode t_mode,
   return CEED_ERROR_SUCCESS;
 }
 
+/**
+  @brief Apply basis evaluation from nodes to arbitrary points
+
+  @param[in]  basis      CeedBasis to evaluate
+  @param[in]  num_points The number of points to apply the basis evaluation to;
+                           the backend will specify the ordering in CeedElemRestrictionCreateBlocked()
+  @param[in]  t_mode    \ref CEED_NOTRANSPOSE to evaluate from nodes to points;
+                          \ref CEED_TRANSPOSE to apply the transpose, mapping from points to nodes
+  @param[in]  eval_mode \ref CEED_EVAL_INTERP to use interpolated values,
+                          \ref CEED_EVAL_GRAD to use gradients
+  @param[in]  x_ref    CeedVector holding reference coordinates of each point
+  @param[in]  u        Input CeedVector, of length `num_points * num
+  @param[out] v        Output CeedVector, of length `num_points * num_comp` for `CEED_NOTRANSPOSE` with `CEED_EVAL_INTERP`
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedBasisApplyAtPoints(CeedBasis basis, CeedInt num_points, CeedTransposeMode t_mode, CeedEvalMode eval_mode, CeedVector x_ref, CeedVector u,
+                           CeedVector v) {
+  return CEED_ERROR_SUCCESS;
```

Comment:
I think `u` is in the FE basis (does not depend on `num_points`). We probably don't need `num_points` to be passed explicitly because it's implied by the shape of `x_ref` and `v`.

Reply:
My thought here was that we could have a oversized vector that holds the maximum number of points for any element in the mesh rather than creating appropriately sized vectors for every element

Reply:
That could work, but I don't know how to set a safe upper bound that isn't kinda huge. I'm not too concerned at this stage because we need to avoid premature optimization. At some point we'll have a spatial tree structure for locating points and will want it to run on the GPU. Pragmatically, I think we should just let it run on the CPU (even when doing implicit solves on GPU) through 2023 and plan to assess performance and discuss GPU-friendly data structures after that.

Reply:
Is it premature optimization? We know that creating vectors on GPU backends is expensive because of the amount of JiT involved, and this happens even if only the CPU buffer is used.

To me this mirrors our passing of `num_elem` to `CeedBasisApply` even though we could probably guess the `num_elem` from the vector sizes passed.

Reply:
What JiT is needed to create a vector? (I'm okay with passing this as an extra argument. Perhaps we should make a `ceed/experimental.h` header to make it clear that some new things can be expected to change.)

Reply:
We ran into this with QFunction/Diagonal assembly, though we've changed somewhat how the GPU backends are organized so I'm trying to pin down where we would see that JiT now. It might have changed a bit but we were seeing the kernels for manipulating the data on the device had some sort of creation cost which is why we started caching those vectors.

Reply:
It's not JIT level of cost, but the cost of allocating on device is higher than on CPU. We try to avoid any allocations in the critical path. One thing we could handle simply is make it possible for vector length to be less than allocated length, so that we could have a cheap resize operation.